api/client/build.go

```diff
@@ -72,211 +72,101 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
+	if tempDir != "" {
+		defer os.RemoveAll(tempDir)
+		contextDir = tempDir
 	}
 
-	// If we are doing a build with trusted pulls on "FROM <image>"
-	// instructions, we need to dynamically rewrite these instructions in the
-	// Dockerfile. I (jlhawn) have chosen to do this by re-writing the Tar
-	// stream of the context because it's the only catch-all way of doing it
-	// since there are so many ways that a Dockerfile could be specified but
-	// it always ends up in the context whether it be taken as a Dockerfile
-	// from stdin, a tar archive from stdin, or a file (possibly taken from
-	// outside the context directory with "-f, --file").
-	// TODO: this is awful, so implement a client-side builder that uses the
-	// existing pull logic.
-	if isTrusted() {
-		pipeReader, pipeWriter := io.Pipe()
-		tarReader := tar.NewReader(context)
-
-		go func() {
-			tarWriter := tar.NewWriter(pipeWriter)
-
-			for {
-				hdr, err := tarReader.Next()
-				if err == io.EOF {
-					// Signals end of archive.
-					tarWriter.Close()
-					pipeWriter.Close()
-					return
-				}
-				if err != nil {
-					pipeWriter.CloseWithError(err)
-					return
-				}
+	newDockerfile, err := cli.makeTrustedDockerfileFrom(filepath.Join(contextDir, *dockerfileName))
+	if err != nil {
+		fmt.Errorf("unable to process Dockerfile: %v", err)
+	}
 
-				var content io.Reader = tarReader
-				if hdr.Name == *dockerfileName {
-					// This entry is the Dockerfile. We need to rewrite the
-					// FROM line(s) of the Dockerfile with a resolved trusted
-					// digest reference. This may end up changing the length
-					// of the Dockerfile so we will need to update the tar
-					// header accordingly.
-					dockerfileBuf, err := cli.makeTrustedDockerfileFrom(tarReader, hdr.Size)
-					if err != nil {
-						pipeWriter.CloseWithError(err)
-						return
-					}
-					hdr.Size = int64(dockerfileBuf.Len())
-
-					content = dockerfileBuf
-				}
+	// And canonicalize dockerfile name to a platform-independent one
+	*dockerfileName, err = archive.CanonicalTarNameForPath(*dockerfileName)
+	if err != nil {
+		return fmt.Errorf("cannot canonicalize dockerfile path %s: %v", *dockerfileName, err)
+	}
 
-				if err := tarWriter.WriteHeader(hdr); err != nil {
-					pipeWriter.CloseWithError(err)
-					return
-				}
+	var includes = []string{"."}
 
-				if _, err := io.Copy(tarWriter, content); err != nil {
-					pipeWriter.CloseWithError(err)
-					return
-				}
-			}
+	excludes, err := utils.ReadDockerIgnore(path.Join(contextDir, ".dockerignore"))
+	if err != nil {
+		return err
+	}
 
-		}()
+	// If .dockerignore mentions .dockerignore or the Dockerfile
+	// then make sure we send both files over to the daemon
+	// because Dockerfile is, obviously, needed no matter what, and
+	// .dockerignore is needed to know if either one needs to be
+	// removed.  The deamon will remove them for us, if needed, after it
+	// parses the Dockerfile.
+	keepThem1, _ := fileutils.Matches(".dockerignore", excludes)
```

Comment:
These errors are checked in another call to `ValidateContextDirectory(contextDir, excludes)` below. I'll move that to come before this and mention it in the comment. I also didn't write this code, it's just been unindented.